pkg/cluster/gossip.go

```diff
@@ -0,0 +1,253 @@
+	// Name of the node within the cluster. Must be unique cluster-wide.
+	NodeName string
+
+	// host:port address to advertise to peers to connect to. When unset, the
+	// first discovered IP from AdvertiseInterfaces will be used to find.
+	AdvertiseAddr string
+
+	// Slice of interface names to infer an advertise IP from. Must be set if
+	// AdvertiseAddr is unset.
+	AdvertiseInterfaces flagext.StringSlice
+
+	// List of one or more host:port peer addresses to connect to. Mutually
+	// exclusive with DiscoverPeers.
+	//
+	// If an agent connects to no peers, it will form a one-node cluster until a
+	// peer connects to it explicitly.
+	JoinPeers flagext.StringSlice
+
+	// Discover peers to connect to using go-discover. Mutually exclusive with
+	// JoinPeers.
+	DiscoverPeers string
+
+	// Client pool to use for connecting to peers.
+	Pool *clientpool.Pool
+}
+
+// DefaultGossipConfig holds default GossipConfig options.
+var DefaultGossipConfig = GossipConfig{
+	AdvertiseInterfaces: advertise.DefaultInterfaces,
+}
+
+// ApplyDefaults mutates c with default settings applied. defaultPort is
+// added as the default port for addresses that do not have port numbers
+// assigned.
+//
+// An error will be returned if the configuration is invalid or if an error
+// occurred while applying defaults.
+func (c *GossipConfig) ApplyDefaults(defaultPort int) error {
+	if c.NodeName == "" {
+		hn, err := os.Hostname()
+		if err != nil {
+			return fmt.Errorf("generating node name: %w", err)
+		}
+		c.NodeName = hn
+	}
+
+	if c.AdvertiseAddr == "" {
+		if len(c.AdvertiseInterfaces) == 0 {
+			return fmt.Errorf("one of advertise address or advertise interfaces must be set")
+		}
+
+		addr, err := advertise.FirstAddress(c.AdvertiseInterfaces)
+		if err != nil {
+			return fmt.Errorf("determining advertise address: %w", err)
+		}
+		c.AdvertiseAddr = fmt.Sprintf("%s:%d", addr.String(), defaultPort)
+	} else {
+		c.AdvertiseAddr = appendDefaultPort(c.AdvertiseAddr, defaultPort)
+	}
+
+	if len(c.JoinPeers) > 0 && c.DiscoverPeers != "" {
+		return fmt.Errorf("at most one of join peers and discover peers may be set")
+	} else if c.DiscoverPeers != "" {
+		providers := make(map[string]discover.Provider, len(discover.Providers)+1)
+		for k, v := range discover.Providers {
+			providers[k] = v
+		}
+		// Extra providers used by tests
+		for k, v := range extraDiscoverProviders {
+			providers[k] = v
+		}
+
+		// Custom providers that aren't enabled by default
+		providers["k8s"] = &k8s.Provider{}
+
+		d, err := discover.New(discover.WithProviders(providers))
+		if err != nil {
+			return fmt.Errorf("bootstrapping peer discovery: %w", err)
+		}
+
+		addrs, err := d.Addrs(c.DiscoverPeers, stdlog.New(io.Discard, "", 0)) // TODO(rfratto): log to log.Logger?
+		if err != nil {
+			return fmt.Errorf("discovering peers: %w", err)
+		}
+		c.JoinPeers = addrs
+	}
+
+	for i := range c.JoinPeers {
+		// Default to using the same advertise port as the local node. This may
```

Comment:
Thoughts on making this mandatory?

Reply:
We can't in all cases 😞 The port can be mandatory if you're providing a hard-coded list, but discovered peers from go-discover don't always have ports (they didn't when I used the k8s provider)